staging/src/k8s.io/apimachinery/pkg/runtime/serializer/cbor/internal/modes/decode_test.go

```diff
@@ -45,6 +45,264 @@ func TestDecode(t *testing.T) {
+				if diff := cmp.Diff(e, &cbor.DupMapKeyError{Key: string("hello"), Index: 1}); diff != "" {
+					t.Errorf("diff between actual error and expected error:\n %s", diff)
+				}
+			}),
+		},
+		{
+			name:  "reject two identical indefinite-length text string keys split into chunks differently into map",
+			modes: []cbor.DecMode{modes.DecodeLax},
+			in:    hex("A27F626865636C6C6FFF017F616864656C6C6FFF02"), // {(_ "he", "llo"):1, (_ "h", "ello"):2}
+			into:  map[string]interface{}{},
+			assertOnError: assertOnConcreteError(func(t *testing.T, e *cbor.DupMapKeyError) {
+				if diff := cmp.Diff(e, &cbor.DupMapKeyError{Key: string("hello"), Index: 1}); diff != "" {
+					t.Errorf("diff between actual error and expected error:\n %s", diff)
+				}
+			}),
+		},
+		{
+			name:  "case-sensitive match",
+			modes: []cbor.DecMode{modes.DecodeLax},
+			in:    hex("A2614101614202"), // {"A": 1}
+			into: struct {
+				LowerA int `json:"a"`
+			}{},
+			want: struct {
+				LowerA int `json:"a"`
+			}{
+				LowerA: 0,
+			},
+			assertOnError: assertNilError,
+		},
+		{
+			name:  "case-sensitive match in Decode mode",
+			modes: []cbor.DecMode{modes.Decode},
+			in:    hex("A2614101614202"), // {"A": 1}
+			into: struct {
+				LowerA int `json:"a"`
+			}{},
+			assertOnError: assertOnConcreteError(func(t *testing.T, e *cbor.UnknownFieldError) {
+				if e.Index != 0 {
+					t.Errorf("expected %#v, got %#v", &cbor.UnknownFieldError{Index: 0}, e)
+				}
+			}),
+		},
+		{
+			name:  "exact match before case-insensitive match",
+			modes: []cbor.DecMode{modes.DecodeLax},
+			in:    hex("a2616101614102"), // {"a": 1, "A": 2}
+			into: struct {
+				LowerA int `json:"a"`
+			}{},
+			want: struct {
+				LowerA int `json:"a"`
+			}{
+				LowerA: 1,
+			},
+			assertOnError: assertNilError,
+		},
+		{
+			name:  "exact match before case-insensitive match in Decode mode",
+			modes: []cbor.DecMode{modes.Decode},
+			in:    hex("a2616101614102"), // {"a": 1, "A": 2}
+			into: struct {
+				LowerA int `json:"a"`
+			}{},
+			want: struct {
+				LowerA int `json:"a"`
+			}{
+				LowerA: 1,
+			},
+			assertOnError: assertOnConcreteError(func(t *testing.T, e *cbor.UnknownFieldError) {
+				if e.Index != 1 {
+					t.Errorf("expected %#v, got %#v", &cbor.UnknownFieldError{Index: 1}, e)
+				}
+			}),
+		},
+		{
+			name:  "exact match after case-insensitive match",
+			modes: []cbor.DecMode{modes.DecodeLax},
+			in:    hex("A2614101616102"), // {"A": 1, "a": 2}
+			into: struct {
+				LowerA int `json:"a"`
+			}{},
+			want: struct {
+				LowerA int `json:"a"`
+			}{
+				LowerA: 2,
+			},
+			assertOnError: assertNilError,
+		},
+		{
+			name:  "exact match after case-insensitive match in Decode mode",
+			modes: []cbor.DecMode{modes.Decode},
+			in:    hex("A2614101616102"), // {"A": 1, "a": 2}
+			into: struct {
+				LowerA int `json:"a"`
+			}{},
+			assertOnError: assertOnConcreteError(func(t *testing.T, e *cbor.UnknownFieldError) {
```

Comment:
If you implement the function I suggested earlier it would also be usable in the cases that expect UnknownFieldError.